ocr.ipynb

```diff
@@ -0,0 +1,181 @@
+      "  f\"No `__init__` function found in the class\"\n"
+     ]
+    }
+   ],
+   "source": [
+    "import os\n",
+    "\n",
+    "from forte.data.data_pack import DataPack\n",
+    "from forte.pipeline import Pipeline\n",
+    "from forte.data.data_pack import DataPack\n",
+    "from forte.pipeline import Pipeline\n",
+    "from typing import Any, Iterator\n",
+    "\n",
+    "from ft.onto.payload_ontology import ImagePayload\n",
+    "\n",
+    "from PIL import Image\n",
+    "from forte.data.base_reader import PackReader"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 2,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "class OcrReader(PackReader):\n",
+    "    r\"\"\"\n",
+    "    :class:`StringReader` is designed to read in a list of string variables.\n",
+    "    \"\"\"\n",
+    "\n",
+    "    # pylint: disable=unused-argument\n",
+    "    def _cache_key_function(self, collection) -> str:\n",
+    "        return str(hash(collection)) + \".html\"\n",
+    "\n",
+    "    def _collect(self, image_path) -> Iterator[Any]:  # type: ignore\n",
+    "        r\"\"\"Should be called with param ``Image_directory`` which is a path to a\n",
+    "        folder containing Image files.\n",
+    "\n",
+    "        Args:\n",
+    "            Image_directory: Image directory containing the files.\n",
+    "\n",
+    "        Returns: Iterator over paths to Image files\n",
+    "        \"\"\"\n",
+    "        # construct ImageMeta and store it in DataPack\n",
+    "        return image_path\n",
+    "\n",
+    "    def _parse_pack(self, data_source: str) -> Iterator[DataPack]:\n",
+    "        r\"\"\"Takes a raw string and converts into a :class:`~forte.data.data_pack.DataPack`.\n",
+    "\n",
+    "        Args:\n",
+    "            data_source: str that contains text of a document.\n",
+    "\n",
+    "        Returns: :class:`~forte.data.data_pack.DataPack` containing Document.\n",
+    "        \"\"\"\n",
+    "        pack: DataPack = DataPack()\n",
+    "        payload_idx = 0\n",
+    "        # Read in Image data and store in DataPack\n",
+    "        # add Image payload into DataPack.payloads\n",
+    "        ip = ImagePayload(pack, payload_idx)\n",
+    "        try:\n",
+    "            import pytesseract\n",
+    "        except ModuleNotFoundError as e:\n",
+    "            raise ModuleNotFoundError(\n",
+    "                \"OcrReader requires 'pytesseract' package to be installed.\"\n",
+    "            ) from e\n",
+    "        image_data = Image.open(data_source) # lazy loading\n",
+    "        ip.set_cache(image_data)\n",
+    "        ocr_text = pytesseract.image_to_string(image_data)\n",
+    "        pack.set_text(ocr_text)\n",
+    "        pack.pack_name = data_source\n",
+    "        \n",
+    "        yield pack\n",
```

Comment:
The loading part is not what we desired, for example, it requires one to explicitly set the cache. And I don't even see metadata being used here.

```
pack: DataPack = DataPack()
jpegMeta = JpegMeta()
ImagePayload(pack, url="some_url", meta=jpegMeta) # payload_index default to 0
```

Now it is done! This is what we call `lazy_loading`, it doesn't happen anywhere near the reader, but behind the scene, when the `Payload` is used, it should load the data to cache based on url and jpegMeta, there don't need to be a step where user set the cache